hypothesis-python/src/hypothesis/searchstrategy/strategies.py

```diff
@@ -480,14 +492,17 @@ def calc_label(self):
 
     def do_draw(self, data):
         # type: (ConjectureData) -> Ex
-        n = len(self.element_strategies)
-        assert n > 0
-        if n == 1:
-            return data.draw(self.element_strategies[0])
-
-        i = cu.integer_range(data, 0, n - 1)
-
-        return data.draw(self.element_strategies[i], label=self.branch_labels[i])
+        assert len(self.element_strategies) > 0
+
+        available_strategies = [x for x in self.element_strategies if x.available(data)]
+        n = len(available_strategies)
+        if n == 0:
+            data.mark_invalid()
```

Comment:
We should follow this line with `assert False, "Unreachable"` as a hint for tools like `mypy`.

We'll also need a test case that executes this branch, e.g. with the `one_of` two bundles.

Reply:
Hmm. Thinking further, we should probably just swap this out for drawing from `sampled_from(self.element_strategies).filter(lambda s: s.available(data).flatmap(lambda s: s)`

That will give us a more efficient shrinking pattern thanks to the special cases for filtering a sampled_from, and make mutations have a more localised effect.

Reply:
implemented by ed13408bae31a87b87c09d8c3c67acdd27459e28

However this solution prevent us from specifying the `label` parameter during the draw...